packages/eslint-plugin/src/rules/no-unnecessary-condition.ts

```diff
@@ -182,17 +182,32 @@ export default createRule<Options, MessageId>({
       return checker.isTupleType(nodeType);
     }
 
-    function isArrayIndexExpression(node: TSESTree.Expression): boolean {
+    // Index accesses are exempt from certain checks about nullish-ness, because
+    // the type returned by the checker is not the strictest
+    function isIndexAccessExpression(node: TSESTree.Expression): boolean {
+      if (node.type !== AST_NODE_TYPES.MemberExpression) {
+        return false;
+      }
+      const objType = getNodeType(node.object);
+      if (checker.isTupleType(objType)) {
+        // If indexing a tuple with a literal, the type will be sound
+        return node.property.type !== AST_NODE_TYPES.Literal;
+      }
+      if (node.computed) {
```

Comment:
It's not just computed expressions - it's all expressions! TS specifically just cares about whether or not the index signature access has been guarded or not!

[playground](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true#code/PTAEAEDsHsFVIMYAsCmCDWKAmBJSWUAPbAQQQRQGdKAoGghAGwEMAnFUBaSSgF1GgAjAFYAuUACU00VlgA8fVgEtIAcwA0oSAFcAtoJSsAfAG46Q4QG0A5M2sBdAHS9oAMSXEsACgCUZi46Czm4e2L5m9Ggs7JzcfKDo4ooqqv4iluhOLu6e4UA)

Reply:
@bradzacher If it's not a computed member access, then it necessarily uses the string index signature, which is checked below.